test/parallel/test-trace-events-async-hooks-worker.js

```diff
@@ -38,60 +42,34 @@ const proc = cp.spawnSync(
     })
   });
 
-console.log(proc.signal);
-console.log(proc.stderr.toString());
-assert.strictEqual(proc.status, 0);
+console.log('prcess exit with signal:', proc.signal);
+console.log('process stderr:', proc.stderr.toString());
 
+assert.strictEqual(proc.status, 0);
 assert(fs.existsSync(filename));
 const data = fs.readFileSync(filename, 'utf-8');
 const traces = JSON.parse(data).traceEvents;
-assert(traces.length > 0);
-// V8 trace events should be generated.
-assert(!traces.some((trace) => {
-  if (trace.pid !== proc.pid)
-    return false;
-  if (trace.cat !== 'v8')
-    return false;
-  if (trace.name !== 'V8.ScriptCompiler')
-    return false;
-  return true;
-}));
 
-// C++ async_hooks trace events should be generated.
-assert(traces.some((trace) => {
-  if (trace.pid !== proc.pid)
-    return false;
-  if (trace.cat !== 'node,node.async_hooks')
-    return false;
-  return true;
-}));
-
-// JavaScript async_hooks trace events should be generated.
-assert(traces.some((trace) => {
+function filterTimeoutTraces(trace) {
   if (trace.pid !== proc.pid)
     return false;
   if (trace.cat !== 'node,node.async_hooks')
     return false;
   if (trace.name !== 'Timeout')
     return false;
   return true;
-}));
-
-// Check args in init events
-const initEvents = traces.filter((trace) => {
-  return (trace.ph === 'b' && !trace.name.includes('_CALLBACK'));
-});
+}
 
-for (const trace of initEvents) {
-  if (trace.name === 'MESSAGEPORT' &&
-    trace.args.data.executionAsyncId === 0 &&
-    trace.args.data.triggerAsyncId === 0) {
-    continue;
-  }
-  if (trace.args.data.executionAsyncId > 0 &&
-    trace.args.data.triggerAsyncId > 0) {
-    continue;
+{
+  const timeoutTraces = traces.filter(filterTimeoutTraces);
+  assert.notDeepStrictEqual(timeoutTraces, []);
+  const threads = new Set();
+  for (const trace of timeoutTraces) {
+    threads.add(trace.tid);
   }
-  assert.fail('Unexpected initEvent: ',
-              util.inspect(trace, { depth: Infinity }));
+  assert.notDeepStrictEqual(timeoutTraces, []);
+  // Since tracing is global, enabling tracing in worker threads
+  // will also enable the same categories in the main thread.
+  console.log('Threads with Timeout traces:', threads);
```

Comment:
Not sure if this is the desired behavior, but this is the current behavior. We may want to revisit this when we migrate to perfetto?

Reply:
cc @addaleax @nodejs/workers

Reply:
I’m not sure what the comment is saying here – do we enable categories from the worker thread? We should not really be able to do that as of 5d80ae3acdd81, right? Am I missing something?

Reply:
> Do we enable categories from the worker thread

Yes, that's what the test is doing (it was added in https://github.com/nodejs/node/pull/26062). Since the tracing controller is per-process, we have to either forbid the workers from enabling any new tracing categories (not sure how to do this, diffing the flags during option parsing?), or allow them to affect the main thread.

Reply:
Okay, I just have a hard time figuring out where/how the test does that?

Reply:
@addaleax Right, it's handling `node.async_hooks` update in `setupTraceCategoryState()` with the value obtained from the v8 intrinsics, which is called in the worker thread main script. Maybe we should just not call `setupTraceCategoryState()` in worker threads?

Reply:
> Maybe we should just not call `setupTraceCategoryState()` in worker threads?

Yeah, that sounds like the right thing to do :+1:

Reply:
Uh, wait, I was being silly about what the test is doing, the flag is already enabled from the main thread, `setupTraceCategoryState()` is not handling any new value, it's just handling whatever passed from the main thread (since `--trace-event-categories` is a per-process option, it simply can not get passed to the workers). So, in summary, the descriptions are bogus (facepalm)

Reply:
Okay, I think we’re on the same page then :)